### PR TITLE
Squash edge/panel signal down to fewer controllers

### PR DESCRIPTION
**Pull request by alextoombs**
**Constraints:**

- Each controller can support up to 8 channels
- All edges consume one channel
- Panels may consume more than one channel, depending upon their size
- Symmetry is _desirable_

**Related Changes:**

These helped in the squash endeavor:

- Assign panels first, then edges: https://github.com/titanicsend/LXStudio-TE/pull/1/commits/6061860bc06caebe67d9d7675739e090ba16a855
- Don't assume the right-most controller is the least utilized: https://github.com/titanicsend/LXStudio-TE/pull/1/commits/981b09d79362c6b51b49090b96202cabe220fed3

**PR:**

This PR manually adjusts the assignment of edges/panels to controllers placed on
the model in order to save us wiring and give us more spare controllers. It better
shares load between each controller; see output from script now below:

```
loaded 36 controllers; expected 49
Total assigned channels, from the vertices:
  - vertex 25 has 1 total controllers, 6 total assigned channel(s)
  - vertex 26 has 1 total controllers, 8 total assigned channel(s)
  - vertex 27 has 1 total controllers, 7 total assigned channel(s)
  - vertex 33 has 1 total controllers, 4 total assigned channel(s)
  - vertex 45 has 1 total controllers, 3 total assigned channel(s)
  - vertex 50 has 1 total controllers, 5 total assigned channel(s)
  - vertex 56 has 1 total controllers, 6 total assigned channel(s)
  - vertex 60 has 1 total controllers, 4 total assigned channel(s)
  - vertex 69 has 1 total controllers, 7 total assigned channel(s)
  - vertex 70 has 1 total controllers, 8 total assigned channel(s)
  - vertex 75 has 1 total controllers, 7 total assigned channel(s)
  - vertex 81 has 1 total controllers, 7 total assigned channel(s)
  - vertex 83 has 2 total controllers, 16 total assigned channel(s)
  - vertex 86 has 1 total controllers, 8 total assigned channel(s)
  - vertex 93 has 2 total controllers, 15 total assigned channel(s)
  - vertex 97 has 2 total controllers, 16 total assigned channel(s)
  - vertex 99 has 1 total controllers, 5 total assigned channel(s)
  - vertex 100 has 2 total controllers, 10 total assigned channel(s)
  - vertex 101 has 1 total controllers, 8 total assigned channel(s)
  - vertex 109 has 2 total controllers, 11 total assigned channel(s)
  - vertex 113 has 1 total controllers, 3 total assigned channel(s)
  - vertex 116 has 1 total controllers, 8 total assigned channel(s)
  - vertex 117 has 1 total controllers, 6 total assigned channel(s)
  - vertex 119 has 1 total controllers, 3 total assigned channel(s)
  - vertex 122 has 1 total controllers, 8 total assigned channel(s)
  - vertex 124 has 1 total controllers, 8 total assigned channel(s)
  - vertex 125 has 1 total controllers, 6 total assigned channel(s)
  - vertex 126 has 2 total controllers, 10 total assigned channel(s)
  - vertex 128 has 1 total controllers, 8 total assigned channel(s)
  - vertex 129 has 1 total controllers, 6 total assigned channel(s)
  total assigned channels: 227
Total assigned channels, from the controllers:
  - controller 25-0 has 6 assigned channel(s)
  - controller 26-0 has 8 assigned channel(s)
  - controller 27-0 has 7 assigned channel(s)
  - controller 33-0 has 4 assigned channel(s)
  - controller 45-0 has 3 assigned channel(s)
  - controller 50-0 has 5 assigned channel(s)
  - controller 56-0 has 6 assigned channel(s)
  - controller 60-0 has 4 assigned channel(s)
  - controller 69-0 has 7 assigned channel(s)
  - controller 70-0 has 8 assigned channel(s)
  - controller 75-0 has 7 assigned channel(s)
  - controller 81-0 has 7 assigned channel(s)
  - controller 83-0 has 8 assigned channel(s)
  - controller 83-1 has 8 assigned channel(s)
  - controller 86-0 has 8 assigned channel(s)
  - controller 93-0 has 8 assigned channel(s)
  - controller 93-1 has 7 assigned channel(s)
  - controller 97-0 has 8 assigned channel(s)
  - controller 97-1 has 8 assigned channel(s)
  - controller 99-0 has 5 assigned channel(s)
  - controller 100-0 has 8 assigned channel(s)
  - controller 100-1 has 2 assigned channel(s)
  - controller 101-0 has 8 assigned channel(s)
  - controller 109-0 has 8 assigned channel(s)
  - controller 109-1 has 3 assigned channel(s)
  - controller 113-0 has 3 assigned channel(s)
  - controller 116-0 has 8 assigned channel(s)
  - controller 117-0 has 6 assigned channel(s)
  - controller 119-0 has 3 assigned channel(s)
  - controller 122-0 has 8 assigned channel(s)
  - controller 124-0 has 8 assigned channel(s)
  - controller 125-0 has 6 assigned channel(s)
  - controller 126-0 has 7 assigned channel(s)
  - controller 126-1 has 3 assigned channel(s)
  - controller 128-0 has 8 assigned channel(s)
  - controller 129-0 has 6 assigned channel(s)
  total assigned channels: 227
number of circuits: 235
```

**Effect:**

- We now use 36 controllers instead of 49
- We'll save a lot of time and wiring
- We'll have 14 spares instead of 1
- We don't have symmetry, but we never really had symmetry in assignment anyways

**What to review for:**

- [ ] Did I miss any constraints?
- [ ] Is symmetry actually more desirable than we think it is?
- [ ] Should we keep squashing?
- [ ] Are any of these manual tweaks nonsensical?

### CODE REVIEW

**Comment by raldi** [comment]
I agree with your prioritization; I expect symmetry to be a mere nice-to-have for panel assembly and the ensuing programming tweaks.

File: script/power_distribution/place_junction_boxes.rb [hunk 1]
```diff
@@ -337,7 +337,8 @@ def print_boxes(junction_boxes)
 
 puts "Total assigned channels, from the controllers:"
 total_channels_from_controllers = 0
-controllers.each do |_, controller_group|
+# This sort_by is ugly, but it just sorts by ascending vertex ID.
+controllers.sort_by { |_, c| c.first.id.split('-')[0].to_i }.each do |_, controller_group|
```
Comment (alextoombs) [inline script/power_distribution/place_junction_boxes.rb:341]: this can just be a sort by `controller.vertex.id` turns out, but whatever

**Review by tcahill — APPROVED** [review]
woah, good stuff

**Comment by alextoombs** [comment]
Fed outputs here into ethernet procurement: https://titanicsend.slack.com/archives/C03650K7AE7/p1658371028583909

**Comment by alextoombs** [comment]
Added a new constraint after discussion with Tom, but that raised the controller count by 5. (we're allocating 41)

This is still the _optimistic allocation_ before the presentation of the problems with the runs from the solid panels: https://titanicsend.slack.com/archives/C03650K7AE7/p1658376299255709

**Pull request merged**